Overhauls modular computer power interaction

atlantiscze opened the pull request:
- All modular computers can now have a tesla link and therefore run off an APC power. This should make tablets much more attractive, as they are less likely to run out of juice now.
- Tesla link passive charging rate increased (100W -> 250W)
- Computer vendor now also offers intellicard slot hardware piece (499T per piece). Intellicard slot also now fits into tablets, though the power requirements are probably too large for it to be of any use when running only off a battery.

Code-wise:
- Modular computers no longer rely on machinery type to use power. While the method i use may be tiny bit more expensive in terms of processing required, the benefit of all devices supporting APC power connection is much more important. Machinery counterpart lost large portion of it's code, pretty much all it does now is relaying of various method calls to the actual computer item.
- Two methods for power interaction are defined for modular computer base type: apc_power() and battery_power(). Passing a parameter of 0 checks whether this type of power supply is available, larger numbers actually use the power.
- APC power now works even when the computer is in a backpack, pocket, or it's item form (laptop). In other words, your laptop won't run out of battery power when you put it enabled into your backpack, as long as you are physically in an area with APC power.

atlantiscze committed:
Overhauls modular computer power interaction
- All modular computers can now have a tesla link and therefore run off an APC power. This should make tablets much more attractive, as they are less likely to run out of juice now.
- Tesla link passive charging rate increased (100W -> 250W)
- Computer vendor now also offers intellicard slot hardware piece (499T per piece). Intellicard slot also now fits into tablets, though the power requirements are probably too large for it to be of any use when running only off a battery.

Code-wise:
- Modular computers no longer rely on machinery type to use power. While the method i use may be tiny bit more expensive in terms of processing required, the benefit of all devices supporting APC power connection is much more important. Machinery counterpart lost large portion of it's code, pretty much all it does now is relaying of various method calls to the actual computer item.
- Two methods for power interaction are defined for modular computer base type: apc_power() and battery_power(). Passing a parameter of 0 checks whether this type of power supply is available, larger numbers actually use the power.
- APC power now works even when the computer is in a backpack, pocket, or it's item form (laptop). In other words, your laptop won't run out of battery power when you put it enabled into your backpack, as long as you are physically in an area with APC power.

mwerezak commented on code/modules/modular_computers/computers/item/modular_computer.dm:
```diff
@@ -492,33 +495,56 @@
+	apc_powered = FALSE
+	if(!battery_module || !battery_module.check_functionality() || battery_module.battery.charge <= 0)
+		return FALSE
+	if(battery_module.battery.use(power_usage * CELLRATE) || ((power_usage == 0) && battery_module.battery.charge))
```
Call `checked_use()` here.

The `battery_module.battery.charge <= 0` above will also become redundant.

atlantiscze replied:
I actually considered that but checked_use() acts a tiny bit differently.

Current code will use battery down to 0 charge units until last tick, therefore even battery_power(0) will return false (this is used to check whether the computer can be turned on, among other things)

checked_use() would leave small amount of energy in the battery, but checked_use(0) would still go through, the computer would start, only to die next tick.

mwerezak replied:
I'd say that's a problem with modular computers, but fair enough.

Kelenius commented:
Make sure to update `/obj/item/modular_computer/tablet/preset/custom_loadout/cheap`, `/obj/item/modular_computer/tablet/preset/custom_loadout/advanced`, and `/datum/gear_tweak/tablet`.

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2017-01-11-PowerTime

atlantiscze committed:
Adds tesla link to tablet's gear_tweak definition
- Intellicard slot is intentionally not included, as it's probably the most expensive hardware out there.

atlantiscze commented:
Former two were already updated, the last one is updated in the third commit.

Techhead0 reviewed (approved):
Looks fine, but I hope that laptops end up de-obj/machinery-ified at some point.
